Add all supported Anymail Providers

Andrew-Chen-Wang opened the pull request:
[//]: # (Thank you for helping us out: your efforts mean great deal to the project and the community as a whole!)

[//]: # (Before you proceed:)

[//]: # (1. Make sure to add yourself to `CONTRIBUTORS.rst` through this PR provided you're contributing here for the first time)
[//]: # (2. Don't forget to update the `docs/` presuming others would benefit from a concise description of whatever that you're proposing)


## Description

[//]: # (What's it you're proposing?)

Providing options for all Anymail-supported ESP:
SES, Mailgun, Mailjet, Mandrill, Postmark, Sendgrid, SendinBlue, SparkPost, Plain/Vanilla Django-anymail, None.


## Rationale

[//]: # (Why does the project need that?)

The default is Mailgun. Financially, it **_could_** be more wise to use SES if the cloud provider is AWS. Some people prefer other options, too.

Alleviates setup for users choosing option other than mailgun.

Fixes #2140 


## Use case(s) / visualization(s)

[//]: # ("Better to see something once than to hear about it a thousand times.")

Selecting an email provider from the beginning alleviates setup later on.

Andrew-Chen-Wang commented:
I do have an issue, though. I'm very confused with the dashes ("-") where there are curly braces, e.g. {%- if %} in production.py. Could it be clarified somewhere inside the cookiecutter/cookiecutter docs?

browniebroke commented:
The dashes are not specific to `cookiecutter-django`, it's coming from the template engine that the `cookiecutter` CLI uses under the hood, Jinja. You can read more about them in this section of their documentation:

https://jinja.palletsprojects.com/en/2.11.x/templates/#whitespace-control

browniebroke commented:
FYI I've looked into simplifying our test setup in #2470, which should make this change a lot more doable.

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/production.py:
```diff
@@ -182,16 +182,89 @@ class MediaRootGoogleCloudStorage(GoogleCloudStorage):
 # https://anymail.readthedocs.io/en/stable/installation/#anymail-settings-reference
 ANYMAIL = {
+    # https://anymail.readthedocs.io/en/stable/esps/
+    {%- if cookiecutter.mail_service == 'Amazon SES' %}
```
It would _probably_ be more readable for us to limit the amount of `elif cookiecutter.mail_service ==` statements and group `EMAIL_BACKEND` and `ANYMAIL` dict in the same block so that they are next to each other, for instance:

```
{%- if cookiecutter.mail_service == 'Mailgun' %}
# https://anymail.readthedocs.io/en/stable/esps/mailgun/
EMAIL_BACKEND = "anymail.backends.mailgun.EmailBackend"
ANYMAIL = {
    "MAILGUN_API_KEY": env("MAILGUN_API_KEY"),
    "MAILGUN_SENDER_DOMAIN": env("MAILGUN_DOMAIN"),
    "MAILGUN_API_URL": env("MAILGUN_API_URL", default="https://api.mailgun.net/v3"),
}
...
```

That means we'll have to reapeat a bit the `ANYMAIL = {` part, but it will make it easier to follow all the settings for 1 provider. Right now it takes some overhead to think about the output for each provider.

browniebroke commented on cookiecutter.json:
```diff
@@ -33,6 +33,17 @@
+    "Sendgrid",
+    "SendinBlue",
+    "SparkPost",
+    "Plain/Vanilla Django-Anymail"
```
As a user of the template, I'm not sure what "plain/vanilla" implies while it's pretty clear that the others are provider name. Is it going to spin up a local for me and send emails directly from Django? Do I need to set this up myself?

Since the question is for "Mail service", maybe we could label this option as "Other (SMTP)"? I'm open to suggestions if you think of a better name.

browniebroke commented on tests/test_cookiecutter_generation.py:
```diff
@@ -73,9 +73,22 @@ def context():
+    {"mail_service", "Sendgrid"},
+    {"mail_service", "SendinBlue"},
+    {"mail_service", "SparkPost"},
+    {"mail_service", "Other SMTP"},
```
```suggestion
    {"mail_service": "AWS SES"},
    {"mail_service": "Mailgun"},
    {"mail_service": "Mailjet"},
    {"mail_service": "Mandrill"},
    {"mail_service": "Postmark"},
    {"mail_service": "Sendgrid"},
    {"mail_service": "SendinBlue"},
    {"mail_service": "SparkPost"},
    {"mail_service": "Other SMTP"},
```

Andrew-Chen-Wang replied:
Yea, running too tests right now. @browniebroke UNSUPPORTED_COMBINATIONS running into errors with the second one.

Andrew-Chen-Wang replied:
I'm having trouble with one test:
```
UNSUPPORTED_COMBINATIONS = [
    {"cloud_provider": "GCP", "mail_service": "Amazon SES"},
    {"cloud_provider": "None", "use_whitenoise": "n"},
    {"cloud_provider": "None", "mail_service": "Amazon SES"},
]
```
Error:
```

=================================================== FAILURES ====================================================
_________________________________ test_error_if_incompatible[invalid_context1] __________________________________
tests/test_cookiecutter_generation.py:208: in test_error_if_incompatible
    assert result.exit_code != 0
E   assert 0 != 0
E    +  where 0 = <Result /private/var/folders/_x/y10g5kjj0yxbklqsmxmr_3qr0000gq/T/pytest-of-andrewwork/pytest-4/test_error_if_incompatible_inv1/cookies/bake00/my_test_project>.exit_code
```
Coming from this test:
```
@pytest.mark.parametrize("invalid_context", UNSUPPORTED_COMBINATIONS)
def test_error_if_incompatible(cookies, context, invalid_context):
    """It should not generate project an incompatible combination is selected."""
    context.update(invalid_context)
    result = cookies.bake(extra_context=context)

    assert result.exit_code != 0
    assert isinstance(result.exception, FailedHookException)
```

Andrew-Chen-Wang commented:
I'm having trouble with one test:
```
UNSUPPORTED_COMBINATIONS = [
    {"cloud_provider": "GCP", "mail_service": "Amazon SES"},
    {"cloud_provider": "None", "use_whitenoise": "n"},
    {"cloud_provider": "None", "mail_service": "Amazon SES"},
]
```
Error:
```

=================================================== FAILURES ====================================================
_________________________________ test_error_if_incompatible[invalid_context1] __________________________________
tests/test_cookiecutter_generation.py:208: in test_error_if_incompatible
    assert result.exit_code != 0
E   assert 0 != 0
E    +  where 0 = <Result /private/var/folders/_x/y10g5kjj0yxbklqsmxmr_3qr0000gq/T/pytest-of-andrewwork/pytest-4/test_error_if_incompatible_inv1/cookies/bake00/my_test_project>.exit_code
```
Coming from this test:
```
@pytest.mark.parametrize("invalid_context", UNSUPPORTED_COMBINATIONS)
def test_error_if_incompatible(cookies, context, invalid_context):
    """It should not generate project an incompatible combination is selected."""
    context.update(invalid_context)
    result = cookies.bake(extra_context=context)

    assert result.exit_code != 0
    assert isinstance(result.exception, FailedHookException)
```

Andrew-Chen-Wang commented:
Actually, for the unsupported_combinations, should I have the ones I added excluded? The cloud provider being GCP with Amazon SES doesn't make sense, but maybe we could still let it happen?

Andrew-Chen-Wang commented:
I've removed the mail providers unsupported combinations as the aforementioned combination was a pain. Once I changed the `SUPPORTED_COMBINATIONS` to have, for example, `{"cloud_provider": "GCP", "mail_service": "Mailgun"},`, I was running into errors with `{"cloud_provider": "None", "mail_service": "Amazon SES"},` and all cloud provider None + mail service combinations (except for Amazon SES since it's not included). I'm dragging in time at the moment, since the latest commit has most of the testing functionality; the only thing missing is the unsupported combinations for None and GCP with Amazon SES @browniebroke

browniebroke commented:
The unsupported combination is enforced in our pre-generation hook:

https://github.com/pydanny/cookiecutter-django/blob/35d74b6386b566a4494286330161276313f0bd75/hooks/pre_gen_project.py#L63-L70

That's what making the test to pass: the test checks that when an unsupported combination is selected, the project is NOT generated. 

It would be nice to add it, but if you're stretched in time, I'm happy to leave this as is and open a follow up issue to document & fix later.
